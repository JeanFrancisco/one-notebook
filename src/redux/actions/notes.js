import Swal from 'sweetalert2';
import { addDoc, collection, deleteDoc, doc, getDocs, orderBy, updateDoc } from '@firebase/firestore';
import {
    NOTES_NEW_NOTE,
    NOTES_SET_ACTIVE,
    NOTES_DISPLAY_IN_SIDEBAR,
    NOTES_UPDATED_NOTE,
    NOTES_DROP_NOTE,
    NOTES_CLEANING_WHEN_LOGOUT,
    NOTES_CHANGE_ORDER,
} from '../constants/typesNotes';
import { uploadFile } from '../../helpers/files';
import { db as firestoreDB } from '../../config/firebase';

export const actionAddNewNote = () => {
    return async (dispatch, getState) => {
        const { uid: user_uid } = getState().auth;

        const newNote = {
            title: '',
            body: '',
            urlImage: '',
            createdAt: new Date().getTime(),
        };

        const doc = await addDoc( collection(firestoreDB, `${user_uid}/journal/notes`), newNote);

        const createdNote = {
            id: doc.id,
            ...newNote,
        };

        dispatch( actionNewNote( createdNote ) );
        dispatch( actionSetActiveNote( createdNote ) );
    }
};

export const actionNewNote = ( createdNote ) => ({
    type: NOTES_NEW_NOTE,
    payload: createdNote,
});

export const actionSetActiveNote = ( activeNote ) => ({
    type: NOTES_SET_ACTIVE,
    payload: activeNote,
});

export const actionSavingChanges = () => {
    return async (dispatch, getState) => {
        const user_uid = getState().auth.uid;
        const activeNote = getState().notes.active;

        const { id: note_id, ...noteForFirestore } = activeNote;
        await updateDoc( doc(firestoreDB, `${user_uid}/journal/notes/${ note_id }`), noteForFirestore )
            .then(() => {
                dispatch( refreshNote(activeNote) );

                Swal.fire({
                    title: 'Guardada exitosamente',
                    html: '<p><small>Su nota se guardó correctamente</small></p>',
                    icon: 'success',
                    timer: 1300,
                    showConfirmButton: true,
                    confirmButtonText: 'Aceptar',
                    confirmButtonColor: 'limegreen',
                });
            })
            .catch(reason => {
                console.log( reason.code );
            });
    }
}

export const refreshNote = ( updatedNote ) => ({
    type: NOTES_UPDATED_NOTE,
    payload: { ...updatedNote },
});

export const getFirebaseNotes = () => {
    return async ( dispatch, getState ) => {
        const { uid } = getState().auth;

        const snapNotes = await getDocs( collection(firestoreDB, `${uid}/journal/notes`) );
        const notes = [];

        snapNotes.forEach( snapHijo => {
            notes.push({
                id: snapHijo.id,
                ...snapHijo.data(),
            })
        });

        dispatch( displayNotes(notes) );
    }
}

export const displayNotes = ( notes ) => ({
    type: NOTES_DISPLAY_IN_SIDEBAR,
    payload: notes,
});

export const actionSetAscendingOrder = () => ({
    type: NOTES_CHANGE_ORDER,
    payload: {
        order: 'asc',
    }
});

export const actionSetDescendingOrder = () => ({
    type: NOTES_CHANGE_ORDER,
    payload: {
        order: 'desc',
    }
});

export const actionReorderByRule = ( orderBy ) => ({
    type: NOTES_CHANGE_ORDER,
    payload: {
        orderby: orderBy,
    }
});

export const actionStartUploadFile = (file) => {
    return async (dispatch, getState) => {
 
        Swal.fire({
            title: 'Subiendo archivos...',
            text: 'Por favor, espere un momento...',
            allowOutsideClick: false,
            showConfirmButton: false,
            willOpen: () => {
                Swal.showLoading();
            },
        });

        const activeNote = getState().notes.active;
        const { secure_url } = await uploadFile(file);

        dispatch( actionSetActiveNote({ ...activeNote, urlImage: secure_url }) );
        await dispatch( actionSavingChanges() );

        Swal.close();
    }
}

export const actionStartDeletingNote = (note_id) => {
    return async (dispatch, getState) => {
        const user_uid = getState().auth.uid;

        try {
            const resp = await deleteDoc( doc( firestoreDB, `${user_uid}/journal/notes/${note_id}`) );

            dispatch( deleteNote( note_id ) );

            Swal.fire({
                icon: 'success',
                timer: 1250,
                title: 'Eliminado',
                text: 'Su nota fue eliminada',
                confirmButtonText: 'Aceptar',
                confirmButtonColor: 'green',
                showCloseButton: true,
            });

        } catch (e) {
            Swal.fire({
                icon: 'error',
                html: 'Error en la conexión con Firestore DB',
                title: 'Error. <small>No pudimos terminar la eliminación</small>',
                footer: `<small>For more details: ${e.code}</small>`,
                confirmButtonText: 'Entendido',
                confirmButtonColor: 'oragered', 
                showCloseButton: true,
            });

            console.error(e);
        }
    }
}

export const deleteNote = (noteID) => ({
    type: NOTES_DROP_NOTE,
    payload: noteID
});

export const cleanupNotesLogout = () => ({
    type: NOTES_CLEANING_WHEN_LOGOUT
});