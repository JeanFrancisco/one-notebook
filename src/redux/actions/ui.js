import { UI_FINISH_LOADING, UI_START_LOADING } from "../constants/typesUi";

export const actionStartLoading = () => ({
    type: UI_START_LOADING,
});

export const actionFinishLoading = () => ({
    type: UI_FINISH_LOADING,
});