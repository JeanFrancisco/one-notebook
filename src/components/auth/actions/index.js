import {
    createUserWithEmailAndPassword,
    signInWithEmailAndPassword,
    GoogleAuthProvider,
    signInWithPopup,
    updateProfile,
    getAuth,
    signOut,
} from "firebase/auth";
import Swal from 'sweetalert2';
import { app } from "../../../config/firebase";
import { cleanupNotesLogout } from "../../../redux/actions/notes";
import { actionFinishLoading, actionStartLoading } from "../../../redux/actions/ui";
import {
    AUTH_LOGIN,
    AUTH_LOGOUT,
    AUTH_SET_ERROR,
    AUTH_REMOVE_ERROR
} from "../constants";

export const authLogin = (uid, displayName) => ({
    type: AUTH_LOGIN,
    payload: {
        uid,
        displayName,
    }
});

export const authSignInWithGoogle = () => {
    return async (dispatch) => {

        dispatch( actionStartLoading() );

        const auth = getAuth( app );

        await signInWithPopup( auth, new GoogleAuthProvider() )
        .then( userCredentials => {
            // const { user: { displayName, uid } } = userCredentials;
            
            // dispatch( authLogin(uid, displayName) );
        })
        .catch( reason => {
            Swal.fire({
                icon: 'warning',
                html: '<small>No se pudo iniciar la sesión. <br/> <br/> ¿Alcanzó a ver si ocurrió algo: con el Pop-up de autenticación de Google?</small>',
                title: 'Oups! Lo sentimos.',
                iconHtml: '<i class="fa fa-warning"></i>',
                footer: `<small>Reason Code: ${reason.code}</small>`,
                showCloseButton: true,
                confirmButtonText: 'Aceptar',
            });
        })
        .finally(() => {
            dispatch( actionFinishLoading() );
        });
    };
}

export const authEmailPasswordSignIn = ({ email, password }) => {
    return async dispatch => {
        dispatch( actionStartLoading() );

        const auth = getAuth(app);

        await signInWithEmailAndPassword(auth, email, password)
            .then( ({ user }) => {
                // const { uid, displayName } = user;

                // dispatch( authLogin(uid, displayName) );
            }).catch( reason => {
                Swal.fire({
                    title: 'Oups! <small>Algo ocurrió mientras iniciabamos tu sesión.</small>',
                    html: `<p><b><small>¿Qué acciones puedes hacer?</small></b></p>
                        <ul style="text-align: left">
                            <li><small>Puedes revisar por favor tu email y contraseña.</small></li>
                            <li><small>También que te puedes conectar a Internet. En especial a los servicios de Google.</small></li>
                        </ul>`,
                    confirmButtonText: 'Entendido',
                    icon: 'warning',
                    iconColor: 'orange',
                    showCloseButton: true,
                    confirmButtonColor: 'orangered',
                    footer: `<small>Reason Code: ${reason.code}</small>`,
                })
            }).finally(() => {
                dispatch( actionFinishLoading() );
            });
    }
}

export const authEmailPasswordSignUp = ({ email, password, username }) => {
    return async (dispatch) => {
        dispatch( actionStartLoading() );

        const auth = getAuth(app);

        await createUserWithEmailAndPassword(auth, email, password)
            .then( async userCredentials => {

                const { user } = userCredentials;
                await updateProfile(user, { displayName: username });

                // dispatch( authLogin(user.uid, user.displayName) );

            }).catch( reason => {
                Swal.fire({
                    title: 'Oups! <small>Ahora mismo no podemos registrarte.</small>',
                    html: '<p>Por favor... <br/> <small>Regresa más tarde y vuelve a intentarlo ;) .</small><p>',
                    icon: 'info',
                    iconColor: 'orange',
                    confirmButtonText: 'Entiendo',
                    confirmButtonColor: 'limegreen',
                    showCloseButton: true,
                    footer: `<small>Reason Code: ${reason.code}</small>`,
                });
            }).finally(() => {
               dispatch( actionFinishLoading() );
            });
    }
}

export const authFirebaseLogout = () => {
    return async dispatch => {
        const auth = getAuth(app);

        await signOut(auth);

        dispatch( authLogout() );

        dispatch( cleanupNotesLogout() );
    }
}

export const authLogout = () => ({ type: AUTH_LOGOUT });

export const setAuthError = ( msg ) => ({
    type: AUTH_SET_ERROR,
    payload: msg
});

export const removeAuthError = () => ({ type: AUTH_REMOVE_ERROR });