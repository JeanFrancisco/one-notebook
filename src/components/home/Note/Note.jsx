import PropTypes from 'prop-types';
import NoteAppBar from '../NoteAppBar/NoteAppBar';
import { useForm } from '../../../hooks/useForm';
import { useEffect, useRef } from 'react';
import { useDispatch } from 'react-redux';
import { actionSetActiveNote, actionStartDeletingNote } from '../../../redux/actions/notes';
import defaultImageDemo from '../../../images/83968430_104662571102161_9176297833785982976_n.jpg';

const Note = ({ contentNote: thisNote }) => {

    const dispatch = useDispatch();

    const [formNotes, handleInputChange, , resetForm ] = useForm(thisNote);

    /**
     * 
     console.log( thisNote ); // thisNote sí cambia. Pero formNotes no lo hace, porque el hook useForm no se vuelve a reiniciar.
     console.log( formNotes );  // ¡ Reported Bug... formNotes: Mantiene el estado inicial de la primera vez que se renderizó el componente.

     Para solucionar este problema.
     Nosotros reseteamos el formulario, con la función de reset del useForm hook.
     Pero sin crear un ciclo infinito, ya que podría producirse al hacer dependencias con
     el hook useEffect(() => {}, [ thisNote ]); ó también, tal como se probó:
     No se actualize porque siempre toma el valor inicial de la thisNote, sin actualizar los cambios en 
     el formValues o en el state interno del hook useForm(). Para solucionar esto último agregamos el useRef() hook.
    */

    const initialNoteID = useRef( thisNote.id );

    useEffect(() => {

        if( thisNote.id !== initialNoteID.current ) {
            resetForm( thisNote );
            initialNoteID.current = thisNote.id;
        }

    }, [thisNote]);

    // El useRef() + useEffect() anteriores, cumplen la misma función que el siguiente IF:
    // if(thisNote.id !== formNotes.id ) {
    //     resetForm( thisNote );
    // }

    useEffect(() => {

        dispatch( actionSetActiveNote(formNotes) );

    }, [formNotes, dispatch]);

    const imgUrl = thisNote.urlImage || defaultImageDemo;

    const handleDeleteNote = () => {
        dispatch( actionStartDeletingNote(thisNote.id) );
    }

    return <div className="note">
        <NoteAppBar />

        <form className="note__form">
            <input
                type="text"
                name="title"
                value={ formNotes.title }
                onChange={ handleInputChange }
                placeholder="Type an awesome title"
                className="note__input-title animate__animated animate__slideInUp animate__slow"
                />

            <textarea
                name="body"
                value={ formNotes.body }
                onChange={ handleInputChange }
                placeholder="Empty"
                className="note__textarea animate__animated animate__fadeIn animate__slow"
                ></textarea>
        </form>

        <div className="note__footer">
            <img
                src={ imgUrl }
                alt="[note-img-preview]"
                className="note__image-preview animate__animated animate__fadeIn" />

            <button className="btn note__button-delete"
                onClick={ handleDeleteNote }
                >
                <span className="fas fa-trash-alt fa-sm"></span> &nbsp;
                Eliminar Nota</button>
        </div>
    </div>
}

Note.propTypes = {
    contentNote: PropTypes.shape({
        title: PropTypes.string.isRequired,
        body: PropTypes.string.isRequired,
        createdAt: PropTypes.number.isRequired,
    }).isRequired
};

export default Note;