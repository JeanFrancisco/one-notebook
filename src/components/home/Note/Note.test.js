import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { actionSetActiveNote, actionStartDeletingNote } from "../../../redux/actions/notes";
import Note from "./Note";

jest.mock('../../../redux/actions/notes', () => ({
    actionSetActiveNote: jest.fn(),
    actionStartDeletingNote: jest.fn(),
}));

const mockStore = configureStore([ thunk ]);

const store = mockStore({
    auth: {},
    notes: {
        notes: [{
            id: 'TEST_NOTE_ID',
            body: 'Lorem impsu',
            title: 'An Tests Title',
            createdAt: 0,
        }],
        active: {
            id: 'TEST_NOTE_ID',
            body: 'Lorem impsu',
            title: 'An Tests Title',
            createdAt: 0,
        },
    }
});

store.dispatch = jest.fn();

describe('Pruebas al componente de <Note />.jsx', () => {

    const wrapper = mount(<Provider store={ store }>
        <Note contentNote={ store.getState().notes.active } />
    </Provider>);

    test('Debe hacer match con el snapshot por defecto', () => {
        expect( wrapper ).toMatchSnapshot();
    });

    test('Debe llamarse a la accion actionStartDeletingNote() cuando le de click al bóton de Eliminar la Acción', () => {
        wrapper.find('.note__button-delete.btn').prop('onClick')();

        expect( actionStartDeletingNote ).toHaveBeenCalledWith( store.getState().notes.active.id );
    });

    test('Debe llamarse a la acción de actionSetActiveNote() cuando se actualice los input del formulario', () => {
        const evt = {
            target: {
                name: 'title',
                value: 'Other Title For Testing',
            }
        };

        wrapper.find('input[name="title"]').simulate('change', evt);


        /*
            wrapper.find('input[name="title"]').prop('onChange')(evt);
         */

        /**
         *  Hacerlo de la forma como se muestra arriba, produce un error
         *  sobre "una actualización del ESTADO o State interno del componente":
            -- Warning: An update to Note inside a test was not wrapped in act(...).
            -- When testing, code that causes React state updates should be wrapped into act(...):
            -- act(() => {})
         *  que debería encerrarse en una funcion act(...)
         **/

        expect( actionSetActiveNote ).toHaveBeenCalledTimes(1);
        expect( actionSetActiveNote ).toHaveBeenCalledWith({
            ...store.getState().notes.active,
            title: evt.target.value
        });

        const new_content_body = 'An Awesome Body Text Change For Tests';

        wrapper.find('textarea').simulate('change', {
            target: {
                name: 'body',
                value: new_content_body,
            }
        });

        expect( actionSetActiveNote ).toHaveBeenCalledWith({
            ...store.getState().notes.active,
            title: evt.target.value,
            body: new_content_body,
        });
    });
});