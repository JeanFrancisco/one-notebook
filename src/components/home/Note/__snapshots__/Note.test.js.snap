// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas al componente de <Note />.jsx Debe hacer match con el snapshot por defecto 1`] = `
<div
  className="note"
>
  <div
    className="note__appbar"
  >
    <span
      className="note__appbar-date"
    >
      28 de agosto de 2020
    </span>
    <input
      id="selector__file-photo"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <button
      className="btn btn-primary"
      onClick={[Function]}
    >
      <span
        className="fas fa-cloud-upload-alt fa-sm"
      >
         
      </span>
        Subir Imagen
    </button>
    <button
      className="btn btn-primary"
      onClick={[Function]}
    >
      <i
        className="fas far fa-save fa-sm"
      />
        Guardar
    </button>
  </div>
  <form
    className="note__form"
  >
    <input
      className="note__input-title animate__animated animate__slideInUp animate__slow"
      name="title"
      onChange={[Function]}
      placeholder="Type an awesome title"
      type="text"
      value="An Tests Title"
    />
    <textarea
      className="note__textarea animate__animated animate__fadeIn animate__slow"
      name="body"
      onChange={[Function]}
      placeholder="Empty"
      value="Lorem impsu"
    />
  </form>
  <div
    className="note__footer"
  >
    <img
      alt="[note-img-preview]"
      className="note__image-preview animate__animated animate__fadeIn"
      src="83968430_104662571102161_9176297833785982976_n.jpg"
    />
    <button
      className="btn note__button-delete"
      onClick={[Function]}
    >
      <span
        className="fas fa-trash-alt fa-sm"
      />
         Eliminar Nota
    </button>
  </div>
</div>
`;
